tests/app/main/views/test_buyers.py

```diff
@@ -61,3 +64,83 @@ def test_should_show_buyers_contact_details(self, data_api_client):
         assert name == "Test Buyer"
         assert email == "test_buyer@example.com"
         assert phone == "02078888888"
+
+
+@mock.patch('app.main.views.buyers.data_api_client')
+class TestAddBuyerDomainsView(LoggedInApplicationTest):
+    @pytest.mark.parametrize("role,expected_code", [
+        ("admin", 200),
+        ("admin-ccs-category", 403),
+        ("admin-ccs-sourcing", 403),
+    ])
+    def test_get_page_should_only_be_accessible_to_specific_user_roles(self, data_api_client, role, expected_code):
+        self.user_role = role
+        response = self.client.get('/admin/buyers/add-buyer-domains')
+        actual_code = response.status_code
+        assert actual_code == expected_code, "Unexpected response {} for role {}".format(response.status_code, role)
+
+    @pytest.mark.parametrize("role,expected_code", [
+        ("admin", 302),
+        ("admin-ccs-category", 403),
+        ("admin-ccs-sourcing", 403)
+    ])
+    def test_post_page_should_only_be_accessible_to_specific_user_roles(self, data_api_client, role, expected_code):
+        self.user_role = role
+        response = self.client.post('/admin/buyers/add-buyer-domains',
+                                    data={
+                                        'new_buyer_domain': 'something.org.uk',
+                                    }
+                                    )
+        actual_code = response.status_code
+        assert actual_code == expected_code, "Unexpected response {} for role {}".format(response.status_code, role)
+
+    def test_admin_user_can_add_a_new_buyer_domain(self, data_api_client):
+        response = self.client.post('/admin/buyers/add-buyer-domains',
+                                    data={'new_buyer_domain': 'kev.uk'},
+                                    follow_redirects=True
+                                    )
+        assert response.status_code == 200
+        assert data_api_client.create_buyer_email_domain.call_args_list == [mock.call("kev.uk", "test@example.com")]
+        assert "You’ve added kev.uk" in response.get_data(as_text=True)
```

Comment:
In the Briefs FE we use `assert_flashes()` (a helper method on `BaseApplicationTest`) to check the right message key and category has been sent.  It's not on the Admin FE `BaseApplicationTest` but we could easily copy it over? Or ... put it in the poor neglected `test_utils` repo (https://github.com/alphagov/digitalmarketplace-test-utils) for all apps to use?

Reply:
I looked at using this, and yes probably should for this test.  It was breaking where I was doing the weird thing of using flashes for the error messages to render in the template (rather than for a redirect).  I'll patch it in here.